src/main/java/bio/terra/cli/service/GoogleOauth.java

```diff
@@ -142,98 +143,246 @@ public void browse(String url) {
+
+    private IdCredentialListener idCredentialListener;
+    private GoogleAuthorizationCodeFlow googleAuthorizationCodeFlow;
+
+    /**
+     * Private ctor called by create() method to create and associate idCredentialListener and
+     * googleAuthorizationCodeFlow members.
+     */
+    private TerraAuthenticationHelper(
+        List<String> scopes, GoogleClientSecrets clientSecrets, File dataStoreDir)
+        throws IOException, GeneralSecurityException {
+
+      // get a pointer to the credential datastore
+      FileDataStoreFactory fileDataStoreFactory = new FileDataStoreFactory(dataStoreDir);
+
+      // create an IdCredentialListener pointing at the credential datastore in order to store and
+      // retrieve ID tokens
+      idCredentialListener =
+          new IdCredentialListener(
+              fileDataStoreFactory, StoredCredential.DEFAULT_DATA_STORE_ID, ID_TOKEN_STORE_KEY);
+
+      // create the code flow object, pointing at the credential datastore in order to store and
+      // retrieve ID tokens, and registering our IdCredentialListener for callbacks on token
+      // creation/refresh
+      googleAuthorizationCodeFlow =
+          new GoogleAuthorizationCodeFlow.Builder(
+                  GoogleNetHttpTransport.newTrustedTransport(), JSON_FACTORY, clientSecrets, scopes)
+              .setDataStoreFactory(fileDataStoreFactory)
+              .setAccessType("offline")
+              .setApprovalPrompt("force")
+              .setCredentialCreatedListener(idCredentialListener)
+              .addRefreshListener(idCredentialListener)
+              .build();
+    }
+
+    /**
+     * Create an instance of {@code TerraAuthenticationHelper}, which is used both for initiating an
+     * OAuth browser flow, as well as storing and retrieving credentials in a DataStore.
+     *
+     * @param scopes OAuth scopes to request when launching an OAuth authentication flow
+     * @param clientSecrets Application Client Secrets used to obtain credentials on behalf of CLI
+     * @param dataStoreDir Directory where credentials store lives, used for cred storage/retrieval
+     * @return
+     * @throws IOException
+     * @throws GeneralSecurityException
+     */
+    public static TerraAuthenticationHelper create(
+        List<String> scopes, GoogleClientSecrets clientSecrets, File dataStoreDir)
+        throws IOException, GeneralSecurityException {
+      return new TerraAuthenticationHelper(scopes, clientSecrets, dataStoreDir);
+    }
+
+    public GoogleAuthorizationCodeFlow getGoogleAuthorizationCodeFlow() {
+      return googleAuthorizationCodeFlow;
+    }
+
+    private static <T extends Serializable> void deleteFromDataStore(
+        DataStore<T> dataStore, String key) throws IOException {
+      if (!dataStore.containsKey(key)) {}
+
+      dataStore.delete(key);
+    }
+
+    /** Delete stored ID token from the credential store */
+    public void deleteStoredIdToken() throws IOException {
```

Comment:
Would it make sense to have these objects implement `AutoCloseable` so we could delete them in a try-with-resources block?